### ISSUE TITLE
std::size_t warning on 32-bit machines

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#553]
When you compile mlpack on 32-bit machines, you get lots and lots of warnings of this ilk:

```
[  4%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/core/math/columns_to_blocks.cpp.o
cd /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack && /usr/lib/ccache/c++   -DARMA_NO_DEBUG -DBACKTRACE_FOUND -DBACKTRACE_HEADER=\"execinfo.h\" -DBOOST_TEST_DYN_LINK -DNDEBUG -Dmlpack_EXPORTS -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m32 -march=i686 -mtune=atom -fasynchronous-unwind-tables  -std=c++0x -Wall -Wextra -ftemplate-depth=1000 -O3 -fopenmp -fPIC -I/home/ryan/rpmbuild/BUILD/mlpack-2.0.1 -I/home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/..    -o CMakeFiles/mlpack.dir/core/math/columns_to_blocks.cpp.o -c /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/core/math/columns_to_blocks.cpp
In file included from /usr/include/armadillo:48:0,
                 from /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/../mlpack/core/arma_extend/arma_extend.hpp:33,
                 from /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/../mlpack/prereqs.hpp:76,
                 from /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/../mlpack/core.hpp:196,
                 from /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/core/dists/gaussian_distribution.hpp:18,
                 from /home/ryan/rpmbuild/BUILD/mlpack-2.0.1/src/mlpack/core/dists/gaussian_distribution.cpp:15:
/usr/include/armadillo_bits/compiler_setup.hpp:87:100: note: #pragma message: WARNING: disabled use of 64 bit integers, as std::size_t is smaller than 64 bits
       #pragma message ("WARNING: disabled use of 64 bit integers, as std::size_t is smaller than 64 bits")
```

We should try and figure out what is triggering these warnings and whether or not we can disable it, so that we can have a less noisy 32-bit build.  These warnings aren't an issue---mlpack doesn't use any 64-bit types if it's on a 32-bit build.

**Comment by ankurayadav** [comment]
I would like to work on this problem but I have a 64 bit machine. I think I will try using a virtual machine.

**Comment by rcurtin** [comment]
Yes, a VM should be sufficient to reproduce this problem.

**Comment by ankurayadav** [comment]
I just build mlpack on ubuntu 14.04 32bit but didn't get any build warning.
Please check [https://gist.github.com/ankurayadav/2b4fdc0b99f3a86defd9](https://gist.github.com/ankurayadav/2b4fdc0b99f3a86defd9).

**Comment by na1taneja2821** [comment]
@rcurtin I saw this issue today and tried to summarize, the problem in build warnings. As can be seen from various online sources the Armadillo library when compiled with the C++ 11 sets the default word size to 64bit. The exact code mentioned in the Armadillo package can be found here: http://fossies.org/linux/armadillo/include/armadillo_bits/compiler_setup.hpp.
Another, thing is that the armadillo library is only being used with arma_extend.hpp, which is used to build the cxx objects. Since, we cannot change the distribution code of armadillo, I suggest that we should implement a check to find whether the machine is 32 Bit or 64 Bit in the arma_extend.hpp. Depending upon that we can simply define ARMA_32BIT_WORD and not define ARMA_USE_U64S64. I have checked by removing the ARMA_USE_U64S64 part and including ARMA_32BIT_WORD in the file, which gives a clean build.
If the solution is fine, I can go ahead with the implementation.

**Comment by rcurtin** [comment]
na1taneja2821: have you reproduced the warnings on your setup?

**Comment by na1taneja2821** [comment]
@rcurtin Yes, I am using Ubuntu 12.04 on a 32-bit virtual machine. Those warnings came up while building the project.

**Comment by na1taneja2821** [comment]
Currently, I am thinking of having a check in CMake files to check if the system is 32 bit or 64 bit. If it is 32bit we can have an option -D ARMA_32BIT_WORD during compilation. But, I am unable to find the code that generates the flags set during compilation.

**Comment by rcurtin** [comment]
Fixed in #564.
